### ISSUE TITLE
intersectBed specific error: line number ... of file ... has ... fields, but ... were expected

### ISSUE THREAD

**Issue by chrarnold** [arq5x/bedtools2#643]
I am receiving a reproducible error intersectBed for a particular gzipped bed file that appears to be completely fine, there are no issues with it at first sight:

1) decompressing works totally fine both via gunzip as well as zless or similar measures
2) Within R, I can directly read in the compressed file via read_tsv without any issues
3) I can use bedtools sort on the file and it runs through just fine. Afterwards, bedtools intersectBed on the sorted file also works fine, so it must be an issue specific to intersectBed
4) I had this already a year ago or so, and now again, it seems to be a rare phenomenon that only occurs in particular circumstances. Other files that have been generated in the same way are fine.

However, when running the following command:

`intersectBed  -a 1.bed.gz -b 1.bed.gz >test.gz`, 
I receive:
`Error: line number 2011585 of file 1.bed.gz has 8 fields, but 10 were expected.`

This is quite a tricky bug that affects my automated pipelines, so I am wondering whether someone has a clue? I put the file here so you can take a look for yourself:
https://www.dropbox.com/s/4d86tp5h9dyqkhm/1.bed.gz?dl=0

**Comment by jmarshall** [comment]
Thanks for the test file. This is a fairly spectacular buffering bug that triggers when EOF is seen on the compressed stream when the zlib decompression code is in a particular state. Apparently this state is very uncommon, so the bug is only triggered on a very few files.

The following fixes it (at present `realloc()` returns `NULL` (on some operating systems…) when `have` is 0 and hence the exception is spuriously thrown):

```diff
diff --git a/src/utils/general/InflateStreamBuf.h b/src/utils/general/InflateStreamBuf.h
index 7c37339d..4790d702 100644
--- a/src/utils/general/InflateStreamBuf.h
+++ b/src/utils/general/InflateStreamBuf.h
@@ -99,7 +99,9 @@ public:
            }           

            unsigned int have = GZBUFSIZ - strm.avail_out;
-           buffer=(char*)std::realloc(buffer,sizeof(char)*(total+have));
+           if (have > 0) {
+               buffer=(char*)std::realloc(buffer,sizeof(char)*(total+have));
+           }
            if(buffer==NULL) {
                throw ("out of memory");
            }
```

(Although if I was going to fix _inflateStreamBuf.h_ fully, I'd remove the memory allocation entirely, as there's no need for code at this low level to be made complicated by doing that.)

**Comment by chrarnold** [comment]
Hi, wow, that sounds super rare indeed, I had a feeling this must be a bug since I had this only twice or so within the last year or so, having produced hundreds or probably more likely thousands of files like this.

I am curious about the specific corner case - why is it so spectacular, could you elaborate a bit why this is so rare?

Thanks for taking a look, and fixing it. I suspect it will be integrated into the next release?

**Comment by chrarnold** [comment]
What is the status here, actually, has a fix been integrated into the latest release already?

**Comment by IanSudbery** [comment]
Yeah, I'd be greatful to know this also.

**Comment by jmarshall** [comment]
As it happens, I was simultaneously updating this fix to be a bit less basic! That's now rebased on top of the fix as pushed — PR #684 if it's of any interest.
